#### Q1. Consider the following sensor readings: temperature = np.array([25, 28, 31, 35, 38, 27, 33, 40]) 
#### A sensor has an average measurement error of +2°C. Perform the following without using a for loop: 

##### a. Add 2°C to every reading using vectorization. 

In [2]:
import numpy as np
temperature = np.array([25, 28, 31, 35, 38, 27, 33, 40])
temperature_corrected = temperature + 2
print(temperature_corrected)

[27 30 33 37 40 29 35 42]


##### b. Convert all temperatures from Celsius to Fahrenheit

In [3]:
fahrenheit = (9/5) * temperature_corrected + 32
print(fahrenheit)

[ 80.6  86.   91.4  98.6 104.   84.2  95.  107.6]


##### c. Identify all readings greater than 32°C using Boolean indexing. 

In [4]:
readings = temperature_corrected[temperature_corrected > 32]
print(readings)

[33 37 40 35 42]


##### d. Count how many readings exceed 32°C. 

In [5]:
count = np.sum(temperature_corrected > 32)
print(count)

5


##### e. Explain how vectorization and Boolean indexing can make data processing more efficient than explicitly iterating through every value. 

Vectorization makes data processing efficient by performing operations on an entire array at once instead of using loops to process each element individually. This reduces execution time and makes the code shorter and easier to understand.

Boolean indexing allows us to filter and select only the elements that satisfy a given condition. For example, it can directly select all temperature readings greater than 32°C without checking each value separately.

#### Q2. The following matrix represents the number of daily steps recorded by four users over three days: steps = np.array([ [5000, 6200, 7100], [8000, 7500, 9000], [4500, 5100, 4800], [9000, 8500, 9500]]) 
#### Perform the following: 

##### a. Calculate the total steps recorded. 

In [6]:
steps = np.array([ [5000, 6200, 7100], [8000, 7500, 9000], [4500, 5100, 4800], [9000, 8500, 9500]])
total = np.sum(steps)
print(total)

84200


##### b. Calculate the mean number of steps. 

In [7]:
print(np.mean(steps))

7016.666666666667


##### c. Find the maximum and minimum values. 

In [8]:
print(np.max(steps))
print(np.min(steps))

9500
4500


##### d. Calculate the total steps for each day using axis=0. 

In [9]:
print(np.sum(steps, axis=0))

[26500 27300 30400]


##### e. Calculate the total steps for each user using axis=1. 

In [10]:
print(np.sum(steps, axis=1))

[18300 24500 14400 27000]


##### f. Find the user/day position containing the maximum number of steps using argmax()

In [11]:
print(np.unravel_index(np.argmax(steps), steps.shape))

(np.int64(3), np.int64(2))


#### Q3. Write a Python program using NumPy to perform the following operations: 

##### a. Create the following NumPy array: original = np.array([1, 2, 3, 4, 5, 6]) 

In [12]:
original = np.array([1, 2, 3, 4, 5, 6])

##### b. Create a slice containing elements from index 1 to 4 and store it in a variable named subset. 

In [13]:
subset = original[1:4]
print(subset)

[2 3 4]


##### c. Modify the first element of subset to 999 and display both original and subset. 

In [14]:
subset[0] = 999
print(original)
print(subset)

[  1 999   3   4   5   6]
[999   3   4]


##### d. Create another slice from original, but this time use .copy(). Modify its first element to 500 and display original and the copied array. 

In [15]:
copied = original[1:4].copy()
copied[0] = 500
print(original)
print(copied)

[  1 999   3   4   5   6]
[500   3   4]


##### e. Create a NumPy array containing the numbers from 1 to 12 using np.arange() and reshape it into a 3 × 4 matrix. 

In [16]:
arr = np.arange(1, 13)
matrix = arr.reshape(3, 4)
print(matrix)

[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]]


##### f. From the 3 × 4 matrix, extract the following using NumPy indexing and slicing: • First row  • Last row  • Second column  • Elements from rows 1–2 and columns 2–3  

In [17]:
matrix[0]

array([1, 2, 3, 4])

In [18]:
matrix[-1]

array([ 9, 10, 11, 12])

In [19]:
matrix[:, 1]

array([ 2,  6, 10])

In [20]:
matrix[0:2, 1:3]

array([[2, 3],
       [6, 7]])

##### g. Create a flattened version of the 3 × 4 matrix using both flatten() and ravel(). 

In [21]:
flat = matrix.flatten()
rav = matrix.ravel()

print(flat)
print(rav)

[ 1  2  3  4  5  6  7  8  9 10 11 12]
[ 1  2  3  4  5  6  7  8  9 10 11 12]


##### h. Modify an element in the array returned by ravel() and observe whether the original array is affected. 

In [22]:
rav[0] = 100
print(rav)
print(matrix)

[100   2   3   4   5   6   7   8   9  10  11  12]
[[100   2   3   4]
 [  5   6   7   8]
 [  9  10  11  12]]


##### i. Modify an element in the array returned by flatten() and observe whether the original array is affected. 

In [23]:
flat[0] = 200

print(flat)
print(matrix)

[200   2   3   4   5   6   7   8   9  10  11  12]
[[100   2   3   4]
 [  5   6   7   8]
 [  9  10  11  12]]


##### j. Display the shape, ndim, size, and dtype of the original 3 × 4 matrix. 

In [24]:
print(matrix.shape)
print(matrix.ndim)
print(matrix.size)
print(matrix.dtype)


(3, 4)
2
12
int64


#### Q4. A simple cognitive-assistive system wants to predict a user's assistance score based on three factors: • Sleep hours  • Activity level  • Stress level  Consider: y = np.array([40, 65, 30, 85]) X = np.array([ [6, 70, 3], [5, 50, 6], [8, 80, 2], [4, 30, 8] ]) Here, each row represents one user and each column represents a feature. Perform the following: 

##### a. Display the shape and dimensions of X. 

In [27]:
import numpy as np
X = np.array([
    [6, 70, 3],
    [5, 50, 6],
    [8, 80, 2],
    [4, 30, 8]
])
y = np.array([40, 65, 30, 85])
print(X.shape)
print(X.ndim)

(4, 3)
2


##### b. Calculate: X.T and explain what the transpose represents.

In [28]:
XT = X.T
print(XT)

[[ 6  5  8  4]
 [70 50 80 30]
 [ 3  6  2  8]]


##### c. Calculate the matrix product: X.T @ X 

In [29]:
product = X.T @ X
print(product)

[[  141  1430    96]
 [ 1430 14700   910]
 [   96   910   113]]


##### d. Calculate the matrix inverse using: np.linalg.inv() where appropriate. 

In [30]:
inverse = np.linalg.inv(X.T @ X)
print(inverse)

[[ 1.10521428 -0.09848746 -0.14581398]
 [-0.09848746  0.00891203  0.01190129]
 [-0.14581398  0.01190129  0.0368847 ]]


##### e. Implement the Ordinary Least Squares equation: 𝛽 = (𝑋𝑋) 𝑋𝑦 using NumPy.

In [31]:
beta = np.linalg.inv(X.T @ X) @ X.T @ y
print(beta)

[-0.15921454  0.14083853 10.06302242]


##### f. Explain what the resulting coefficients represent in the context of the three user features. 

The coefficients represent the effect of each feature on the assistance score, while keeping the other features constant.

Sleep hours coefficient = -0.1592: A 1-unit increase in sleep hours decreases the predicted assistance score by about 0.1592. 
Activity level coefficient = 0.1408: A 1-unit increase in activity level increases the predicted assistance score by about 0.1408. 
Stress level coefficient = 10.0630: A 1-unit increase in stress level increases the predicted assistance score by about 10.0630. 

Thus, the coefficients indicate how strongly each of the three features influences the predicted assistance score.

##### g. A new user has: new_user = np.array([5, 40, 7]) Use your calculated coefficients to obtain a predicted assistance score. 

In [32]:
new_user = np.array([5, 40, 7])
prediction = new_user @ beta
print(prediction)

75.27862544779134
